ownCloud/Settings/Passcode/PasscodeViewController.swift

```diff
@@ -0,0 +1,306 @@
+//
+//  PasscodeViewController.swift
+//  ownCloud
+//
+//  Created by Javier Gonzalez on 03/05/2018.
+//  Copyright © 2018 ownCloud GmbH. All rights reserved.
+//
+
+import UIKit
+import ownCloudSDK
+
+typealias PasscodeHandler = (() -> Void)
+
+let numberDigitsPasscode = 4
+let passcodeKeychainAccount = "passcode-keychain-account"
+let passcodeKeychainPath = "passcode-keychain-path"
+
+// MARK: - Interface view mode
+enum PasscodeInterfaceMode {
+    case addPasscodeFirstStep
+    case addPasscodeSecondStep
+    case unlockPasscode
+    case unlockPasscodeError
+    case deletePasscode
+    case deletePasscodeError
+    case addPasscodeFirstSetpAfterErrorOnSecond
+}
+
+class PasscodeViewController: UIViewController, Themeable {
+
+    // MARK: - Handler
+    var handler: PasscodeHandler?
+
+    // MARK: - Overlay
+    var passcodeFromFirstStep: String?
+    var passcodeMode: PasscodeInterfaceMode?
+    var hiddenOverlay: Bool?
+
+    // MARK: - Overlay view
+    @IBOutlet var overlayView: UIView!
+    @IBOutlet var logoTVGView : VectorImageView!
+
+    // MARK: - Messages and input text
+    @IBOutlet weak var messageLabel: UILabel?
+    @IBOutlet weak var errorMessageLabel: UILabel?
+    @IBOutlet weak var passcodeValueTextField: UITextField?
+
+    // MARK: - Buttons
+    @IBOutlet weak var number0Button: ThemeButton?
+    @IBOutlet weak var number1Button: ThemeButton?
+    @IBOutlet weak var number2Button: ThemeButton?
+    @IBOutlet weak var number3Button: ThemeButton?
+    @IBOutlet weak var number4Button: ThemeButton?
+    @IBOutlet weak var number5Button: ThemeButton?
+    @IBOutlet weak var number6Button: ThemeButton?
+    @IBOutlet weak var number7Button: ThemeButton?
+    @IBOutlet weak var number8Button: ThemeButton?
+    @IBOutlet weak var number9Button: ThemeButton?
+
+    @IBOutlet weak var cancelButton: ThemeButton?
+
+    // MARK: - Initalization view
+    init(mode: PasscodeInterfaceMode, hiddenOverlay: Bool?, handler: PasscodeHandler?) {
+        super.init(nibName: "PasscodeViewController", bundle: nil)
+        self.passcodeFromFirstStep = nil
+        self.passcodeMode = mode
+        self.hiddenOverlay = hiddenOverlay
+        self.handler = handler
+    }
+
+    required init?(coder aDecoder: NSCoder) {
+        fatalError("init(coder:) has not been implemented")
+    }
+
+    override func viewWillAppear(_ animated: Bool) {
+        super.viewWillAppear(animated)
+
+        Theme.shared.register(client: self)
+    }
+
+    override func viewDidAppear(_ animated: Bool) {
+        super .viewDidAppear(animated)
+    }
+
+    override func viewWillDisappear(_ animated: Bool) {
+        super.viewWillDisappear(animated)
+
+        Theme.shared.unregister(client: self)
+    }
+
+    override func viewDidLoad() {
+        super.viewDidLoad()
+
+        self.loadUI()
+        UIDevice.current.setValue(UIInterfaceOrientation.portrait.rawValue, forKey: "orientation")
```

Comment:
How then? All the examples that I can found about how to rotate a device programatically use this approach

Reply:
Looks like it is not necessary to force the rotation to portrait. Using this the view it is presented always in portrait:
`
    // MARK: - Rotation Control

    override var shouldAutorotate: Bool {
        return false
    }

    override var supportedInterfaceOrientations: UIInterfaceOrientationMask {
        return .portrait
    }
`

Reply:
Unfortunately, this does not solve the basic problem that only one UIViewController can be presented by a UIViewController at a time:
- present this on a UIViewController that already presents one => fails, no lock screen, no protection
- present this on a UIViewController that doesn't present one => shows lock screen, but has potential to break the app if that view controller wants to present a view controller while the lock VC is already presented

In addition, as you noticed, you get warnings / errors in the log from UIKit when trying to present that view controller right after starting the app.

Why not use the solution I already sketched out for you above?

> You can leverage these to:
> - create a new UIWindow
> - use the passcode view controller as its root view controller
> - respond to PasscodeViewController.supportedInterfaceOrientations with portrait only 
> - respond to -[UIApplicationDelegate application:supportedInterfaceOrientationsForWindow:] with PortraitOnly for that window, so it's always shown in portrait mode (f.ex. by subclassing UIWindow as LockWindow, use LockWindow above, then check if the window passed into this method is of class LockWindow, and, if it is, return a mask that only contains portrait mode)
> 
> What you get with this is:
> - a window that completely covers the main window
> - a way to display the passcode view controller always in portrait mode, regardless of which orientation the device is currently in and existing view controllers currently display in

Plus, I might add: this causes no side effects with other view controllers.

Reply:
I see... I wanted to try to use the viewController because with it I can show the animation when the `PasscodeViewController` is presented but you are right doing that I could have those problems.

I will change it to use directly a subclass of Window as you say 👍